sympy/matrices/dense.py

```diff
@@ -16,22 +17,343 @@
 from sympy.utilities.misc import filldedent
 from sympy.utilities.decorator import doctest_depends_on
 
-from sympy.matrices.matrices import (MatrixBase,
-    ShapeError, a2idx, classof)
+from sympy.matrices.matrices import (MatrixBase, MatrixError,
+    ShapeError, CommonMatrix, a2idx, classof)
+
+
+def _force_mutable(x):
+    """Return a matrix as a Matrix, otherwise return x."""
+    if getattr(x, 'is_Matrix', False):
+        return x.as_mutable()
+    elif isinstance(x, Basic):
+        return x
+    elif hasattr(x, '__array__'):
+        a = x.__array__()
+        if len(a.shape) == 0:
+            return sympify(a)
+        return Matrix(x)
+    return x
 
 
 def _iszero(x):
     """Returns True if x is zero."""
     return x.is_zero
 
 
-class DenseMatrix(MatrixBase):
+class InplaceMatrix(object):
+    """A Matrix-like object where all operations happen
+    in place.  This is meant to be used inside Matrix
+    classes and should not be used directly. It has
```

Comment:
By starting it with `_` ?  Currently it is not exported so for someone to get their hands on an `InplaceMatrix` they would have to do `from sympy.matrices.dense import InplaceMatrix`.  Is that private enough?